doc/spinel-protocol-src/spinel-basis-intro-encoding.md

```diff
@@ -0,0 +1,106 @@
+
+[DBUS]: https://www.freedesktop.org/wiki/Software/dbus/
+
+As with the D-Bus, the Spinel data packing format also defines a terse modeling language for describing the format of data packed for interchange between the OS and the NCP. With Spinel, however, the modeling language is an optional notational convenience, mainly of use in protocol definitions. As most NCP programming environments are highly constrained, it is not necessary to implement a structured validating parser for Spinel packed data to implement the Spinel protocol.
+
+Goals:
+
+ *  Be lightweight and favor direct representation of values.
+ *  Support lists and structures.
+ *  Allow fields to be appended to structures while maintaining backward compatibility.
+ *  Use an easily readable and memorable language for data modeling.
+
+The Spinel data packing format is a method of encoding and decoding several "primitive" types of scalar data, e.g. fixed width integers, fixed-size network addresses, et cetera, and some simple aggregate types, i.e. arrays of a specific type, structures with fields of varying type, Unicode text strings.
+
+Each primitive datatype has an ASCII character associated with it in the corresponding modeling language. Fields in structures are identified by their position. The data model for any encoding of the Spinel data packing format can be represented as a strings of modeling language characters characters. These strings are called "type signatures" and some examples follow:
+
+ *  `C`: An unsigned integer encoded as a single octet.
+ *  `C6U`: An unsigned integer encoded as a single octet, followed by a 128-but IPv6 address, followed by a Unicode text string.
+ 
+In each case, the data is represented exactly as described. For example, an array of 10 IPv6 address is stored as 160 octets.
+
+## Primitive Types
+
+Char | Name                 | Description
+-----|:---------------------|:------------------------------
+ `.` | DATATYPE_VOID        | Empty data type. Used internally.
+ `b` | DATATYPE_BOOL        | Boolean value. Encoded in 8-bits as either 0x00 or 0x01. All other values are illegal.
+ `C` | DATATYPE_UINT8       | Unsigned 8-bit integer.
+ `c` | DATATYPE_INT8        | Signed 8-bit integer.
+ `S` | DATATYPE_UINT16      | Unsigned 16-bit integer.
+ `s` | DATATYPE_INT16       | Signed 16-bit integer.
+ `L` | DATATYPE_UINT32      | Unsigned 32-bit integer.
+ `l` | DATATYPE_INT32       | Signed 32-bit integer.
+ `i` | DATATYPE_UINT_PACKED | Packed Unsigned Integer. See (#packed-unsigned-integer).
+ `6` | DATATYPE_IPv6ADDR    | IPv6 Address. (Big-endian)
+ `E` | DATATYPE_EUI64       | EUI-64 Address. (Big-endian)
+ `e` | DATATYPE_EUI48       | EUI-48 Address. (Big-endian)
+ `D` | DATATYPE_DATA        | Arbitrary data. See (#data-blobs).
+ `d` | DATATYPE_DATA_WLEN   | Arbitrary data with prepended length. See (#data-blobs).
+ `U` | DATATYPE_UTF8        | A text string encoded in [modified UTF-8][MUTF8] and terminated by U+0000 NUL character.
```

Comment:
I just noticed this change. Changing this from UTF-8 to MUTF8 is a big change that is fundamentally incompatible with the current spinel parser, and seems unnecessary (especially replacing the null termination!). Why is this a good idea?